### ISSUE TITLE
Read OrcFile error when schema in Orc file and the table file don't consist

### ISSUE THREAD

**Issue by liujp** [apache/incubator-gluten#5638]
### Backend

VL (Velox)

### Bug description

when the schema in Orc file is different from the table schema(especially the field name), the gluten read error. as follows:
schema in ORC file:
<img width="2080" alt="image" src="https://github.com/apache/incubator-gluten/assets/11934146/2222f056-130a-42cd-b7e0-8ee150e6ce95">
 the user table is:
CREATE TABLE parquet_test
(
    treatment TINYINT,
    numerator DOUBLE,
    denominator TINYINT,
    numerator_pre BIGINT,
    denominator_pre TINYINT,
    Y DOUBLE,
    X1 INT,
    X2 INT,
    X3 INT,
    X3_string STRING,
    X7_needcut BIGINT,
    X8_needcut BIGINT,
    weight DOUBLE,
    distance DOUBLE
)

then execute the query: 
val df = spark.sql( "SELECT count(Y) from udf_table")
a null value returns,

![](https://file.tapd.woa.com/compress/compress_img/700/tapd_70077997_base64_1715042626_348.png?src=/tfl/captures/2024-05/tapd_70077997_base64_1715042626_348.png)

but use query: val df = spark.sql( "SELECT count(_col5) from udf_table")
return ok;
<img width="561" alt="image" src="https://github.com/apache/incubator-gluten/assets/11934146/114172ed-31f5-4abf-aec3-b9abb573b030">

### Spark version

None

**Comment by kecookier** [comment]
Low version hive-orc file not contain field name, only with field index. I will try to fix it.

**Comment by liujp** [comment]
> Low version hive-orc file not contain field name, only with field index. I will try to fix it.

you are right, mainly compact for lower version. thx!

**Comment by liujp** [comment]
> Low version hive-orc file not contain field name, only with field index. I will try to fix it.

velox has already support using different schema(table schema and file schema )，user need to add a schema to create hive table handle
<img width="1101" alt="image" src="https://github.com/apache/incubator-gluten/assets/11934146/94abe518-e76b-4390-8acf-5e43c6956822">

**Comment by kecookier** [comment]
https://github.com/apache/incubator-gluten/blob/main/cpp/velox/substrait/SubstraitToVeloxPlan.cc#L1094

**Comment by kecookier** [comment]
Yes, we can pass the table schema parameter (dataColumns) when creating a HiveTableHandle, but currently it's set to nullptr.

https://github.com/apache/incubator-gluten/blob/main/cpp/velox/substrait/SubstraitToVeloxPlan.cc#L1129    
https://github.com/apache/incubator-gluten/blob/main/cpp/velox/substrait/SubstraitToVeloxPlan.cc#L1094

**Comment by wenwj0** [comment]
@liujp @kecookier Any solution for this? I also met this problem.

**Comment by liujp** [comment]
> @liujp @kecookier Any solution for this? I also met this problem.

Now I add a schema table to ReadRelNode, and serialize it to Protobuf `LocalFiles`. gluten deserialize the schema in LocalFiles, and the pass the schema to HiveTableHandle. But for ORC file, this schema is only the top level, for the columns which has child type, this schema is not unabridged. For this condition, ORC reader should reuse the parquet reader.

**Comment by kecookier** [comment]
@ccat3z

**Comment by beliefer** [comment]
What's the current progress ?

**Comment by zhouyuan** [comment]
@ccat3z Hi, it looks like there is a pending patch for Velox (https://github.com/apache/incubator-gluten/pull/8862) is this the right patch to fix the issue?

**Comment by ccat3z** [comment]
> [@ccat3z](https://github.com/ccat3z) Hi, it looks like there is a pending patch for Velox ([#8862](https://github.com/apache/incubator-gluten/pull/8862)) is this the right patch to fix the issue?

Yes, I will rebase this patch when I'm free.
